### PR TITLE
Trace through eager checkpoint under non-strict tracing

### PR DESCRIPTION
**Pull request by tugsbayasgalan**
Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* #179105
* __->__ #179043

When the non-strict tracing flag is set, make `_is_compiling` return
False so that `torch.utils.checkpoint` takes the eager code path
instead of the compile path. This lets make_fx trace through the full
forward+backward including checkpoint recomputation.

The eager checkpoint machinery (saved tensor hooks, RNG state
stashing) executes correctly during tracing, but get/set_rng_state
are silently dropped from the FX graph. This means dropout inside
checkpointed regions will produce different masks during replay — a
known limitation documented in the tests.

Authored with Claude.

cc @voznesenskym @penguinwu @EikanWang @jgong5 @Guobing-Chen @XiaobingSuper @zhuhaozhe @blzheng @wenzhe-nrv @jiayisunx @kadeng @chauhang @amjames @Lucaskabela @jataylo @azahed98

### CODE REVIEW

**Review by soulitzer — APPROVED** [review]

File: torch/utils/checkpoint.py [hunk 1]
```diff
@@ -1334,7 +1334,9 @@ def __torch_dispatch__(self, func, types, args=(), kwargs=None):
         if isinstance(policy, bool):
             policy = _policy_from_bool(policy)
 
-        is_compiling = _is_compiling(func, args, kwargs)
+        # TODO: eventually we will only rely on tagging for the compile path
+        # and remove the eager checkpoint machinery entirely in compile path.
+        is_compiling = _is_compiling(func, args, kwargs) and not torch.compiler._is_non_strict_tracing()
```
Comment (soulitzer) [inline torch/utils/checkpoint.py]: Why did you move the torch.compiler._is_non_strict_tracing() check out of _is_compiling? Was someone else depending on _is_compiling, and we broke them?

Comment (tugsbayasgalan) [reply torch/utils/checkpoint.py]: I just wanted to be more explicit here.
